edge/consumer.go

```diff
@@ -0,0 +1,69 @@
+package edge
+
+type Consumer struct {
+	edge Edge
+	r    Receiver
+}
+
+func NewConsumer(edge Edge, r Receiver) *Consumer {
+	return &Consumer{
+		edge: edge,
+		r:    r,
+	}
+}
+
+func (ec *Consumer) Run() error {
+	for msg, ok := ec.edge.Next(); ok; msg, ok = ec.edge.Next() {
+		switch typ := msg.Type(); typ {
+		case BeginBatch:
```

Comment:
My only concern, and I'm not sure that it's even valid, is that switch for every message will end up being expensive.

My concern comes from a note in [go-lua](https://github.com/Shopify/go-lua#benchmarks)

>This exercises more of the bytecode interpreter’s inner loop. Here we see the performance impact of Go’s switch implementation. Both go-lua and gopher-lua are an order of magnitude slower than the C Lua interpreter.

I'm not exactly sure what to do about it, but my initial though is to do something like you would with a parser, Where we receive a `BeginBatchMessage` and therefore we know that all further messages should be `PointMessage`s, until we receive a `EndBatchMessage`. This does assume that that there are some semantics to the order in which messages are received.